### ISSUE TITLE
Distinguishing can-i-deploy cases between pact existance and version differences

### ISSUE THREAD

**Issue by mattermack** [pact-foundation/pact_broker-client#30]
I am running into a case that I think is different to other issues I see.  #25 is similar but not exactly the same.   The crux is I cannot tell difference between:
- versions of provider and client that have not be tested in this combination (row 3)
- provider and consumer that both exist but have never had a pact between them

I am trying to use the client to determine if a particular combination has never been tested (which would say that the contracts are different) or if there is no pact at all between these two pacticipants.   I half started to write an app to hit the Pact Broker API.  I think this would be key functionality in the decision of "can-i-deploy".   Is there something that gets me close?

|Consumer Name|Consumer Version|Provider Name|Provider Version |Success|Output|Notes|Existing Exit Code|
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
|PactPhpExampleClient|0.0.30|PactExampleApi|0.1.57|Yes|All verification results are published and successful| |0|
|PactPhpExampleClient|0.0.36|PactExampleApi|1.1.71|No|One or more verifications have failed| |1|
|PactPhpExampleClient|0.0.36|PactExampleApi|1.1.57|Unknown|No results matched the given query|An older version of pact exists but says unknown|1|
|PactPhpExampleClient|0.0.36|ExistantApiButNoneWithPactPhpExampleClient|Latest|Unknown|No results matched the given query|No pact ever existed|1|

**Comment by mattermack** [comment]
Structured.   I am working on some proof of concepts with Helm and k8s.   Pact Broker integration, imo, is key to validated Helm charts before moving to Stage / pre-Prod.   

By being able to differentiate between Row 3 and Row 4, I can determine:
- Row 3 _likely_ does not work together as we know the contract / pact changed
- Row 4 is still ambiguous on what to do.  I could see cases where this is allowed but depends on how Helm charts are analyzed.   

In short, we need _some_ flag to distinguish between the same.   I really need to learn Ruby to contribute :)

**Comment by bethesque** [comment]
Help me understand scenario 4 better. I don't understand why two services that don't actually have a pact between them are being queried with can-i-deploy. Are you specifically naming each pacticipant in the can-i-deploy call? What makes you take that approach, rather than letting the broker just work out the integrations?

eg. It looks like you're doing `can-i-deploy --pacticipant PactPhpExampleClient --version  0.0.36 --pacticipant PactExampleApi` instead of `can-i-deploy --pacticipant PactPhpExampleClient --version  0.0.36`

**Comment by mattermack** [comment]
Yes, I *think* that can work (one could argue the feature is still interesting but that's besides the point).  I made a more complicated scenario to review the possibilities.

|Row|Consumer Name|Consumer Version|Provider Name|Provider Version|Success|
| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |
|1|GorillaClient|0.0.30|ZooApi|0.1.57|Yes|
|2|GorillaClient|0.0.31|ZooApi|0.1.57|Yes|
|3|GorillaClient|0.0.31|ZooApi|1.1.71|No|
|4|GorillaClient|0.0.36|ZooApi|1.1.71|Yes|
|5|GorillaClient|0.0.36|ZooApi|1.1.57|Unknown|
|6|GorillaClient|0.0.36|NoneExistantApi|Latest|Unknown|
|7|ZooApi|0.1.57|StoreApi|2.0.0|Yes|
|8|ZooApi|1.1.71|StoreApi|2.0.0|Yes|
|9|SharkClient|2.0.0|ZooApi|1.1.71|No|
|10|SharkClient|2.0.0|ZooApi|1.1.90|Yes|

As mentioned, I'm exploring tying this into Helm charts and k8s.    I think I can fit these two basic use cases.


## Deployment Use Case 1: the Helm Chart says ZooApi 0.1.57 
`pact-broker.bat can-i-deploy --pacticipant ZooApi  --version 0.1.57 --broker-base-url http://mybroker/ --output table`

If the cluster only has GorillaClient and ZooApi, then we are ok.

| CONSUMER      | C.VERSION | PROVIDER | P.VERSION | SUCCESS? |
| --------------|-----------|----------|-----------|--------- |
| GorillaClient | 0.0.31    | ZooApi   | 0.1.57    | true |
| GorillaClient | 0.0.30    | ZooApi   | 0.1.57    | true |
| ZooApi        | 0.1.57    | StoreApi | 2.0.0     | true |

However, what happens if there are other apps in the Helm chart that have not been tested?  How would we know the delta in the Helm chart from an app-to-app perspective?

## Deployment Use Case 2: the Helm Chart says ZooApi 0.1.71
`pact-broker.bat can-i-deploy --pacticipant ZooApi  --version 0.1.71 --broker-base-url http://mybroker/ --output table` 

If the Helm chart has a Shark Client, we are explicitly in trouble and should not deploy.

| CONSUMER      | C.VERSION | PROVIDER | P.VERSION | SUCCESS? |
| --------------|-----------|----------|-----------|---------
| GorillaClient | 0.0.36    | ZooApi   | 0.1.71    | true |
| GorillaClient | 0.0.31    | ZooApi   | 0.1.71    | false |
| GorillaClient | 0.0.30    | ZooApi   | 0.1.71    | false |
| SharkClient   | 2.0.0     | ZooApi   | 0.1.71    | false |
| ZooApi        | 0.1.71    | StoreApi | 2.0.0     | true |

The same holds true as Deployment 1.  What happens if there are other apps in the Helm chart that have not been tested?  How would we know not to compare ComputerClient to ZooApi?

Noodling on this one a little bit.   Not sure the right answer on how to incorporate Pact Broker into k8s & Helm yet.

**Comment by bethesque** [comment]
Thanks for your response Charles, but I'm still confused. When you give can-i-deploy only one pacticipant, it works out which other pacticipants it has a pact with, and only returns rows for those integrations. You can never get a row for an pair that doesn't have a pact. This is the way I would recommend the tool is used. Is there something about that approach that doesn't work for you?

**Comment by bethesque** [comment]
Did you get any further with this @mattermack?

**Comment by bethesque** [comment]
Closing due to inactivity, but feel free to open if you have more questions @mattermack
